### PR TITLE
[com_fields] SQL - FIND_IN_SET() portable other than mysql

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #12636.
### Summary of Changes
- com_fields  introduced the mysql dialect `FIND_IN_SET()`  function
- added postgresql portabilty
### Testing Instructions

com_fields should work not only  on mysql, still lacks an equivalent function for MSSQL

### CODE REVIEW

File: libraries/joomla/database/query/mysqli.php [hunk 1]
```diff
@@ -141,4 +141,25 @@ public function Rand()
...
+	 *
+	 * @param   string  $value  The value to search for.
+	 *
+	 * @param   string  $set    The set of values.
+	 *
+	 * @return  string  Returns the find_in_set() Mysql translation.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function find_in_set($value, $set)
```
Comment (mbabker) [inline libraries/joomla/database/query/mysqli.php]: Two things:

1) Please rename the method to `findInSet` to match code style rules (we don't use snake case method names)
2) Can you add a method to the base `JDatabaseQuery` class that just returns an empty string?  This way there aren't fatal errors using other drivers.

**Comment by alikon** [comment]
- method renamed from `find_in_set()` to `findInSet()`
- method added to the base class

**Comment by laoneo** [comment]
I have not tested this item.<br /><br />Tested on a mysql environment.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on a024973af53c6c6f2a03fce85c0c0d06be5e6817<br /><br />Tested on a mysql environment. Thanks!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by csthomas** [comment]
Does joomla need new query method?
Why not replace `find_in_set(x, column_with_list)` 
by something more common like `CONCAT(',', x, ',') LIKE CONCAT('%,', column_with_list, ',%')`?

**Comment by alikon** [comment]
i'm not a big fan of this `FIELD_IN_SET()`.... but as it is used by core com_fields......

**Comment by csthomas** [comment]
Then I have prepared a list of possibilities.
There is a few sql string function that can replace `find_in_set`:
- common: LIKE operator
- mysql: locate, position, instr
- postgreSQL: position
- sqlsrv: charindex
- sqlite: instr

**Comment by laoneo** [comment]
@csthomas I can't really remember, but there was a case where the like operation didn't work. I think it was when there was a comma in the search string or in the column_with_list.

**Comment by alikon** [comment]
I don't think adding a new method is a problem, the only problem I see is that this pr don't still provide a MSSQL translation , @csthomas can you test/write the translation maybe using charindex() on MSSQL ? I don't have a MSSQL environment available...

**Comment by csthomas** [comment]
ok

**Comment by alikon** [comment]
now we have the MSSQL translation too  
thanks @csthomas

**Comment by laoneo** [comment]
Does that mean it is tested by @csthomas  on MSSQL?

**Comment by csthomas** [comment]
I have not tested com_fields. 
I tested only `echo $q->select('*')->from('#__content')->where($q->findInSet('your', 'title'));` on mssql as I described at https://github.com/alikon/joomla-cms/pull/25

**Comment by laoneo** [comment]
You can easily create it with com_fields when you create a few article categories and create a few fields and assign them to the different article categories. When you edit now an article only fields should be shown which belong to the category the article is in.

**Comment by PhilETaylor** [comment]
I have tested this item :red_circle: unsuccessfully on 146d188b4ff5cffccdb674168aef52aea3167644<br /><br />When this PR is applied to 351b778254893d1ebbd734a88e3ad2a33b386aff it results in a massive sql error 

`Unknown column 'a.catid' in 'on clause' SQL=SELECT a.*,l.title AS language_title, l.image AS language_image,uc.name AS editor,ag.title AS access_level,ua.name AS author_name,c.title as category_title, c.access, c.published FROM #__fields AS a LEFT JOIN `#__languages` AS l ON l.lang_code = a.language LEFT JOIN #__users AS uc ON uc.id=a.checked_out LEFT JOIN #__viewlevels AS ag ON ag.id = a.access LEFT JOIN #__users AS ua ON ua.id = a.created_user_id LEFT JOIN #__categories AS c ON c.id = a.catid WHERE a.context = 'com_content.article' AND a.state IN (0, 1) ORDER BY a.ordering asc LIMIT 20 Unknown column 'a.catid' in 'on clause' SQL=SELECT COUNT(*) FROM #__fields AS a LEFT JOIN `#__languages` AS l ON l.lang_code = a.language LEFT JOIN #__users AS uc ON uc.id=a.checked_out LEFT JOIN #__viewlevels AS ag ON ag.id = a.access LEFT JOIN #__users AS ua ON ua.id = a.created_user_id LEFT JOIN #__categories AS c ON c.id = a.catid WHERE a.context = 'com_content.article' AND a.state IN (0, 1)
`


screen shot 2016-12-11 at 15 53 11
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by alikon** [comment]
@PhilETaylor  probably your  test site is not updated see #13019
the #__fields table  don't have a column named catid cause has been removed/changed with group_id

**Comment by PhilETaylor** [comment]
My site was a git clone of the main repo as of today

**Comment by zero-24** [comment]
@philetaylor a clean install of staging?

**Comment by zero-24** [comment]
I did have some similiar issues because of a broken setup :)

**Comment by PhilETaylor** [comment]
I have tested this item :red_circle: unsuccessfully on 146d188b4ff5cffccdb674168aef52aea3167644<br /><br />I tested this on joomla-cms/staging at ec08ae2a015daa68c1d98033b0253afdbf485a4d

I installed from scratch, from ec08ae2a015daa68c1d98033b0253afdbf485a4d, and then installed PatchTester Extension, and applied the Patch from PR #12642 

I then went to Admin -> Content -> Fields

I then get that massive SQL issue <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by PhilETaylor** [comment]
Also after applying this patch on the frontend I see: 
screen shot 2016-12-12 at 12 33 39

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 146d188b4ff5cffccdb674168aef52aea3167644<br /><br />Tested on a mysql environment. Thanks!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by laoneo** [comment]
@PhilETaylor I couldn't reproduce your error. Looks like there is some old code around in your installation as the catid got removed with PR #13019. I know you have been asked already, but are you sure you have the latest staging code and all fully synced?

**Comment by PhilETaylor** [comment]
Im sorry but this is impossible. I know how to make a clean slate and I know I can replicate this issue over and over.

**Comment by laoneo** [comment]
In the file administrator/components/com_fields/models/fields.php after line 293 can you add the following code

    echo $query;die;

what is the output?

**Comment by PhilETaylor** [comment]
I have just tried this again with joomla-cms/staging#102d90f65446e57d3f90445e85be87521d8ed325

Applied Patch for PR 12642 with patchtester extension from @mbabker 

Same result 
screen shot 2016-12-13 at 15 21 41

the part of the code changed (note the black dots in the gutter) near line 293
screen shot 2016-12-13 at 15 22 35

The dumped query, dumped before the `$search = $this->getState('filter.search');` line:

`SELECT a.*,l.title AS language_title, l.image AS language_image,uc.name AS editor,ag.title AS access_level,ua.name AS author_name,c.title as category_title, c.access, c.published FROM #__fields AS a LEFT JOIN `#__languages` AS l ON l.lang_code = a.language LEFT JOIN #__users AS uc ON uc.id=a.checked_out LEFT JOIN #__viewlevels AS ag ON ag.id = a.access LEFT JOIN #__users AS ua ON ua.id = a.created_user_id LEFT JOIN #__categories AS c ON c.id = a.catid WHERE a.context = 'com_content.article' AND a.state IN (0, 1)


`

**Comment by mbabker** [comment]
This would mean the remote branch is out of sync then because the patch tester copies the files from the remote branch and replaces them, not patches files (too complex to be able to do that).

**Comment by PhilETaylor** [comment]
@mbabker define "remote branch" in this context? as you can see Im starting with a clean folder checkout from the copy and pasted url of the github repo. See my terminal screen.

**Comment by mbabker** [comment]
The remote branch in the context of a pull request is the branch in the user's fork the pull request comes from.

Generally, if one of the files in the pull request changed after they created that branch and made their changes, it can cause conflicts and unintended side effects like what you're seeing.

**Comment by laoneo** [comment]
Thanks @mbabker for clarification. Will be hard without you in the future.

**Comment by PhilETaylor** [comment]
ok give me an hour and I'll manually prepare a patch file from the PR and apply that with command line patch tools :)

**Comment by mbabker** [comment]
```sh
curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/12642.diff | git apply
```

**Comment by PhilETaylor** [comment]
ok doing it that way and it tests successfully 

I guess today I learned that the PatchTester extension has limitations :)

**Comment by laoneo** [comment]
@waader do you have time to test it on MSSQL as well?

**Comment by laoneo** [comment]
@PhilETaylor great. Can you please hit the successful test button it on the issue tracker https://issues.joomla.org/tracker/joomla-cms/12642

**Comment by PhilETaylor** [comment]
I have tested this item :white_check_mark: successfully on 146d188b4ff5cffccdb674168aef52aea3167644<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12642">issues.joomla.org/tracker/joomla-cms/12642</a>.</sub>

**Comment by laoneo** [comment]
@alikon I assume you tested it on postgres, did you had also a look on MSSQL. I would not merge it before we don't have a MSSQL test.

**Comment by alikon** [comment]
I've tested on posgresql, mysql but not on MSSQL, and yes I agree we need
to wait for someonelse tests as for our rules on pr

**Comment by waader** [comment]
I will test on mssql today or tomorrow when i find some time.

**Comment by csthomas** [comment]
Tested successful on mysql but I can not test this PR on mssql.

After I install joomla from git with mssql I got in backend warning:
"Error loading component: com_fields, Component not found."

**Comment by laoneo** [comment]
@csthomas probably you need to install it from here https://github.com/alikon/joomla-cms/archive/patch-85.zip, as it contains the installer script for MSSQL.

**Comment by waader** [comment]
I tried testing this one with the installation patch applied on mssql. Adding a new category I get 

8155 [Microsoft][SQL Server Native Client 11.0][SQL Server]No column name was specified for column 1 of 'A'. SQL=SELECT * FROM ( SELECT count(id) , ROW_NUMBER() OVER (ORDER BY (select 0)) AS RowNumber FROM #__usergroups) A WHERE A.RowNumber BETWEEN 1 AND 1 

Same when adding a field group
8155 [Microsoft][SQL Server Native Client 11.0][SQL Server]No column name was specified for column 1 of 'A'. SQL=SELECT * FROM ( SELECT count(id) , ROW_NUMBER() OVER (ORDER BY (select 0)) AS RowNumber FROM #__usergroups) A WHERE A.RowNumber BETWEEN 1 AND 1 

I am stucked.

**Comment by alikon** [comment]
i'm quite sorry to say that this should be another different MSSQL issue
`
SELECT * FROM ( SELECT count(id) , ROW_NUMBER() OVER (ORDER BY (select 0)) AS RowNumber FROM #__usergroups) A WHERE A.RowNumber BETWEEN 1 AND 1
`

seems very ugly SQL

**Comment by csthomas** [comment]
What I done:
```
git clone https://github.com/csthomas/joomla-cms.git # it is joomla/joomla-cms.git HEAD~2 for now

curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/12642.diff | git apply
curl https://patch-diff.githubusercontent.com/raw/joomla/joomla-cms/pull/12660.diff | git apply

# GO TO libraries/cms/helper/usergroups.php:197:                                ->select('count(id)')
# AND CHANGE TO
libraries/cms/helper/usergroups.php:197:                                ->select('count(id) AS count')
```

Next I can go to create a new field but when I save I got:

Save failed with the following error: [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Cannot insert the value NULL into column 'fieldparams', table 'joomla.dbo.#__fields'; column does not allow nulls. INSERT fails. SQL=INSERT INTO [#__fields] ([context],[group_id],[assigned_cat_ids],[title],[alias],[label],[default_value],[type],[note],[description],[state],[required],[params],[language],[created_time],[created_user_id],[access]) VALUES ('com_content.article','','76','Field recipes 1','field-recipes-1','Field recipes 1','','text','','','1','0','{"hint":"","image":"","image_alt":"","render_class":"","class":"","disabled":"0","readonly":"0","show_on":"","display":"-1"}','*','2016-12-15 20:12:53','968','1')

**Comment by csthomas** [comment]
@alikon count(id) has to have an alias.

We should fix it in  libraries/cms/helper/usergroups.php or in JDatabaseQuerySqlsrv?

**Comment by alikon** [comment]
yes seems unrelated to `field_in_set()`

**Comment by csthomas** [comment]
@alikon Can I ask you to merge that PR with latest joomla staging.
It would be easier to test that PR without curl...

**Comment by wilsonge** [comment]
OK As this is another step forward I'm going to merge this. And I have opened https://github.com/joomla/joomla-cms/issues/13263 to track the MSSQL testing which looks like it relies on other code to be merged

**Pull request merged**